Fix for Pane controller issue#1670

dtailor90 opened the pull request:
1. Added ability for users to add the following parameters to the url:

        a. hideTree = true
        b. hideInspector = true

   These params will enable users to hide the tree and/or inspector. 
   Params will automatically be removed once the PaneController state is triggered. 

2. Changed the way PaneController state is controlled when opening a new tab. Previously we were only checking if new tab is opened, now (newTabOpened=true) param is added to url to act as a trigger to hide Tree and Inspector by default.

larkin commented on platform/commonUI/browse/res/templates/browse.html:
```diff
@@ -24,7 +24,7 @@
     <div class="abs holder holder-main browse-area s-browse-area browse-wrapper"
          ng-controller="PaneController as modelPaneTree"
-         ng-class="modelPaneTree.visible() ? 'pane-tree-showing' : 'pane-tree-hidden'">
+         ng-class="modelPaneTree.visible() ? 'pane-tree-showing' : 'pane-tree-hidden'" alias="Tree">
```
Hrm, `alias="Tree"` doesn't clearly communicate what this does. 

How about something like `hide-parameter="hideTree"`

larkin commented on platform/commonUI/browse/src/PaneController.js:
```diff
@@ -31,12 +31,20 @@ define(
-            // Fast and cheap: if this has been opened in a new window, hide panes by default
-            this.state = !$window.opener;
+            var alias = $attrs.alias;
+            var aliasParams = $location.search()["hide" + alias];
```
This should be conditional, then you don't need comments and it's more clear what it's doing.

Also, a hasOwnProperty check will check if the parameter is defined, instead of parsing the truthiness of it, which could be confusing.

e.g.

```javascript
if ($attrs.hideParameter && location.search().hasOwnProperty($attrs.hideParameter)) { 
    this.state = false; 
    $location.search(hideParam, undefined);
} else { 
    this.state = true; 
}
```

larkin commented on platform/commonUI/browse/src/windowing/NewTabAction.js:
```diff
@@ -38,6 +38,7 @@ define(
 
             this.urlService = urlService;
             this.open = function () {
+                arguments[0] += "&tabOpened=true"; // add tabOpened param
```
instead of `tabOpened`, use `hideInspector` and `hideTree`.  Don't need the complexity of all three.

This comment is literally duplicating what the code does, so unnecessary.  Let code speak for itself.

larkin commented on platform/commonUI/browse/test/InspectorPaneControllerSpec.js:
```diff
@@ -71,7 +73,12 @@ define(
+                mockLocation = jasmine.createSpyObj("$location", ["search"]);
+                mockLocation.search.andReturn({});
+
+                mockAttrs = jasmine.createSpyObj("$attrs", ["alias"]);
```
a jasmine spy object is an object with spy properties: actual functions.  You don't need that for your tests, you can just use a regular object, i.e. `mockAttrs = {};`

larkin commented on platform/commonUI/browse/test/PaneControllerSpec.js:
```diff
@@ -59,6 +63,11 @@ define(
+                mockLocation = jasmine.createSpyObj("$location", ["search"]);
+                mockLocation.search.andReturn({});
+
+                mockAttrs = jasmine.createSpyObj("$attrs", ["alias"]);
```
same note about mockAttrs as before.

dtailor90 commented:
## Author Checklist
1.  Changes address original issue? Y
2.  Unit tests included and/or updated with changes? Updated and Added new tests
3.  Command line build passes? Y
4.  Changes have been smoke-tested? Y

@larkin please review when you get a chance

larkin commented on platform/commonUI/browse/res/templates/browse.html:
```diff
@@ -24,7 +24,7 @@
     <div class="abs holder holder-main browse-area s-browse-area browse-wrapper"
          ng-controller="PaneController as modelPaneTree"
-         ng-class="modelPaneTree.visible() ? 'pane-tree-showing' : 'pane-tree-hidden'">
+         ng-class="modelPaneTree.visible() ? 'pane-tree-showing' : 'pane-tree-hidden'" hideParameter="hideTree">
```
Html attributes are case insensitive, we stick to lowercase.  In order to get this to camelCase in $attrs as `hideParameter`, you should change this to `hide-parameter` in the HTML.

larkin commented on platform/commonUI/browse/src/windowing/NewTabAction.js:
```diff
@@ -38,6 +38,7 @@ define(
 
             this.urlService = urlService;
             this.open = function () {
+                arguments[0] += "&hideTree=true&hideInspector=true"; //default to hideTree and hideInspector
```
remove comment.

larkin commented on platform/commonUI/browse/test/InspectorPaneControllerSpec.js:
```diff
@@ -71,7 +73,14 @@ define(
                 mockDomainObject.getCapability.andReturn(mockStatusCapability);
 
-                controller = new InspectorPaneController(mockScope, mockAgentService, mockWindow, mockNavigationService);
+                mockLocation = { search: function () {
```
This works fine, however you can be more consistent with surrounding code by using a jasmine spyObject.

larkin replied:
For consistency, can you change this to a spy?  Would be more concise too.

larkin commented on platform/commonUI/browse/test/PaneControllerSpec.js:
```diff
@@ -86,6 +97,26 @@ define(
                 expect(controller.visible()).toBeFalsy();
             });
+
+            it("if url includes hideTree parameter then Tree is not visible", function () {
```
`hideTree` and `hideInspector` are specific uses of hideParameter. You should be more specific: you are testing what hideParameter does.  e.g.

```javascript
describe("specifying hideParameter", function () {
    it("hides panel when parameter present in location", ...);
    it("does not hide panel when parameter not present in location", ...);
});
```

dtailor90 commented:
@larkin added the required tests, rebased and squashed commits. Please review when you get a chance.

larkin commented on platform/commonUI/browse/test/PaneControllerSpec.js:
```diff
@@ -59,6 +63,13 @@ define(
                 );
                 mockWindow = jasmine.createSpyObj("$window", ["open"]);
+
+                mockLocation = { search: function () {
```
For consistency, can you change this to a spy?  And then you don't have to redefine mockLocation down on line 103; you only need to change the return value.

larkin commented on ~:
```diff
@@ -0,0 +1,3 @@
+export HOMEBREW_PREFIX=/Users/dtailor/.homebrew
```
This file shouldn't be checked in.

larkin commented:
# Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
